clang/lib/Sema/SemaDecl.cpp

```diff
@@ -9237,6 +9282,22 @@ static FunctionDecl *CreateNewFunctionDecl(Sema &SemaRef, Declarator &D,
   FunctionDecl *NewFD = nullptr;
   bool isInline = D.getDeclSpec().isInlineSpecified();
 
+  ConstexprSpecKind ConstexprKind = D.getDeclSpec().getConstexprSpecifier();
+  if (ConstexprKind == ConstexprSpecKind::Constinit ||
+      (SemaRef.getLangOpts().C23 &&
+       ConstexprKind == ConstexprSpecKind::Constexpr)) {
+
+    if (SemaRef.getLangOpts().C23)
+      SemaRef.Diag(D.getDeclSpec().getConstexprSpecLoc(),
+                   diag::err_c23_constexpr_not_variable);
```

Comment:
> Hmmm, I'm not certain we need to change anything here

Well, without that move for C we silently create `FunctionDecl` and exit whether it has `constexpr` specifier or not.

> it would be more clear to say that a function cannot be constexpr in C instead of saying you declared something constexpr that isn't a variable. WDYT?

I don't seem to have a problem with this particular message, for C++ similar invalid use cases clang says and does the same. But If that will help users, ok.
I also use `err_c23_constexpr_not_variable` message when diagnosing constexpr applied to a type declaration. Should I add more specific message for this case as well?